File: java/twelve-days/src/main/java/TwelveDays.java
class TwelveDays {
    
    private static String[] days = new String[]{"first", "second", "third", "fourth", "fifth", "sixth", "seventh",
        "eighth", "ninth", "tenth", "eleventh", "twelfth"};
    
    private static String[] gifts = new String[]{
        "a Partridge in a Pear Tree.",
        "two Turtle Doves, ",
        "three French Hens, ",
        "four Calling Birds, ",
        "five Gold Rings, ",
        "six Geese-a-Laying, ",
        "seven Swans-a-Swimming, ",
        "eight Maids-a-Milking, ",
        "nine Ladies Dancing, ",
        "ten Lords-a-Leaping, ",
        "eleven Pipers Piping, ",
        "twelve Drummers Drumming, "
    };

    String verse(int verseNumber) {
        StringBuilder out = new StringBuilder("On the " + days[verseNumber - 1]
                + " day of Christmas my true love gave to me: ");
        
        for (int i = verseNumber; i > 0; i--) {
            if (verseNumber != 1 && i == 1)
                out.append("and ");
            
            out.append(gifts[i-1]);
        }

        out.append("\n");
        
        return out.toString();
    }

    String verses(int startVerse, int endVerse) {
        StringBuilder out = new StringBuilder(verse(startVerse));
        for (int i = startVerse + 1; i <= endVerse; i++) {
            out.append("\n");
            out.append(verse(i));
        }
        
        return out.toString();
    }
    
    String sing() {
        return verses(1, 12);
    }
}

File: fsharp/twelve-days/TwelveDays.fs
module TwelveDays

let days =
    [ "first"; "second"; "third"; "fourth"; "fifth"; "sixth"; "seventh"; "eighth"; "ninth"
      "tenth"; "eleventh"; "twelfth" ]

let numbers =
    [ "a"; "two"; "three"; "four"; "five"; "six"; "seven"; "eight"; "nine"; "ten"; "eleven"; "twelve"]

let gifts =
    [ "Partridge in a Pear Tree"
      "Turtle Doves, and "
      "French Hens, "
      "Calling Birds, "
      "Gold Rings, "
      "Geese-a-Laying, "
      "Swans-a-Swimming, "
      "Maids-a-Milking, "
      "Ladies Dancing, "
      "Lords-a-Leaping, "
      "Pipers Piping, "
      "Drummers Drumming, " ]


let stanza day =
    let giftStr =
        [ day-1 .. -1 .. 0 ]    // count down.
        |> List.map (fun dy -> numbers[dy] + " " + gifts[dy])
        |> List.reduce (fun acc nxt -> acc + nxt)

    $"On the {days[day-1]} day of Christmas my true love gave to me: {giftStr}."


let recite start stop =
    [ start .. stop ]
    |> List.map stanza

;;;
